modules/kvformat/kv-parser.c

```diff
@@ -112,6 +112,28 @@ kv_parser_free(LogPipe *s)
   log_parser_free_method(s);
 }
 
+static gboolean
+kv_parser_process_threaded(LogParser *s, LogMessage **pmsg, const LogPathOptions *path_options, const gchar *input, gsize input_len)
+{
+  LogParser *self = (LogParser *)kv_parser_clone(&s->super);
+
+  gboolean ok = kv_parser_process(self, pmsg, path_options, input, input_len);
+
+  kv_parser_free(&self->super);
+  return ok;
+}
+
+static gboolean
+kv_parser_process_init(LogParser *s, LogMessage **pmsg, const LogPathOptions *path_options, const gchar *input, gsize input_len)
+{
+  if (s->super.cfg && s->super.cfg->threaded)
+    s->process = kv_parser_process_threaded;
```

Comment:
there are other ways to initiate threading (the threaded flag on an input driver), so this is not really safe.

And I would try to avoid any conditionals in driver code if at all possible.

Can't we avoid malloc/free by using a local variable when cloning? But you know what, only a small fraction of the input will be parsed by kv-parser(), so for now let's stick to the clone() stuff, always. We'll see if there are bottlenecks in the area. If there is, I would probably use a pool of KVScanner instances, one dedicated to all threads.

So I think the conditions on threaded should just be reverted.